cluster/addons/dns/kube2sky/kube2sky.go

```diff
@@ -199,41 +216,93 @@ func (ks *kube2sky) addDNSUsingEndpoints(subdomain string, e *kapi.Endpoints) er
 	if err := ks.removeDNS(subdomain); err != nil {
 		return err
 	}
-	return ks.generateRecordsForHeadlessService(subdomain, e, svc)
+	return ks.generateRecordsForHeadlessService(subdomain, e, svc, isNewStyleFormat)
 }
 
 func (ks *kube2sky) handleEndpointAdd(obj interface{}) {
 	if e, ok := obj.(*kapi.Endpoints); ok {
 		name := buildDNSNameString(ks.domain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, false) })
 		name = buildDNSNameString(ks.domain, serviceSubdomain, e.Namespace, e.Name)
-		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e) })
+		ks.mutateEtcdOrDie(func() error { return ks.addDNSUsingEndpoints(name, e, true) })
 	}
 }
 
-func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service) error {
+func (ks *kube2sky) generateRecordsForPortalService(subdomain string, service *kapi.Service, isNewStyleFormat bool) error {
+	recordKey := subdomain
+
+	if isNewStyleFormat {
+		recordLabel, err := randomID(5)
+		if err != nil {
+			return err
+		}
+		recordKey = buildDNSNameString(subdomain, recordLabel)
+	}
 	for i := range service.Spec.Ports {
-		b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, service.Spec.Ports[i].Port))
+		port := &service.Spec.Ports[i]
+		b, err := json.Marshal(getSkyMsg(service.Spec.ClusterIP, port.Port))
 		if err != nil {
 			return err
 		}
-		glog.V(2).Infof("Setting DNS record: %v -> %q\n", subdomain, string(b))
-		if err := ks.writeSkyRecord(subdomain, string(b)); err != nil {
+		glog.V(2).Infof("Setting DNS record: %v -> %q, with recordKey: %v\n", subdomain, string(b), recordKey)
+		if err := ks.writeSkyRecord(recordKey, string(b)); err != nil {
 			return err
 		}
+		if isNewStyleFormat {
+			portSegment := buildPortSegmentString(port.Name, port.Protocol)
+			if portSegment != "" {
+				recordLabel, err := randomID(5)
+				if err != nil {
+					return err
+				}
+				err = ks.generateSRVRecord(subdomain, portSegment, recordLabel, port.Port)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func buildPortSegmentString(portName string, portProtocol kapi.Protocol) string {
+	if portName == "" {
+		// we don't create a random name
+		return ""
+	}
+	portNameSegment := fmt.Sprintf("_%s", portName)
+
+	protocolSegment := string(portProtocol)
+	if protocolSegment == "" {
```

Comment:
You should be able to compare portProtocol to "" directly without copying it to a string and then re-copying that in Sprintf.

```
if portProtocol = "" {
    // log
}
protocolSegment = fmt.Sprintf("_%s", strings.ToLower(portProtocol))